### ISSUE TITLE
Signing plugin not applied

### ISSUE THREAD

**Issue by eygraber** [vanniktech/gradle-maven-publish-plugin#228]
In 0.14.2 the following worked:

```
apply plugin: 'org.jetbrains.dokka'
apply plugin: 'com.vanniktech.maven.publish'

mavenPublish {
  releaseSigningEnabled = !getGpgKey().isEmpty()
}

def getGpgKey() {
  return findProperty("signingKey") ?: ""
}

signing {
  def signingKey = getGpgKey()
  if(!signingKey.isEmpty()) {
    useInMemoryPgpKeys(signingKey, "")
  }
}
```

After updating to 0.15.0 I get the following error:

```
Could not find method signing()
```

If I add:

```
apply plugin: 'signing'
```

the build works. This happens regardless of whether `releaseSigningEnabled` is `true` or `false`.

I also get the following error if I don't apply the signing plugin manually:

```
Invalid MavenPublish Configuration. Unable to find variant to publish named release. Try setting the 'androidVariantToPublish' property in the mavenPublish extension object to something that matches the variant that ought to be published.
```

This happens regardless of whether I manually set `androidVariantToPublish` or if I use the default.

**Comment by gabrielittner** [comment]
In case you want to use in memory signing keys, which works great for CI, you can specify them like this instead:
Signing will probably work if you put it in an `afterEvaluate` blog. To setup in memory signing you can also use `signingInMemoryKey` as name for the Gradle property (or `ORG_GRADLE_PROJECT_signingInMemoryKey` as env var), then we should pick it up automatically.

I will take a look at the `Invalid MavenPublish Configuration` error

**Comment by eygraber** [comment]
Thanks!

**Comment by eygraber** [comment]
I tried using `signingInMemoryKey` and now it fails when configuring the `CloseAndReleaseRepositoryTask`.

Here's what my gradle config looks like:

```
apply plugin: 'org.jetbrains.dokka'
apply plugin: 'com.vanniktech.maven.publish'

mavenPublish {
  sonatypeHost = "S01"
}
```

Here are the env vars I have set up:

```
ORG_GRADLE_PROJECT_mavenCentralUsername
ORG_GRADLE_PROJECT_mavenCentralPassword
ORG_GRADLE_PROJECT_mavenCentralRepositoryUsername
ORG_GRADLE_PROJECT_mavenCentralRepositoryPassword
ORG_GRADLE_PROJECT_signingKey
ORG_GRADLE_PROJECT_signingInMemoryKey
```

And here's the error:

```
> Task :debug-db-base:build

> Task :debug-db-base:dokkaHtml
Initializing plugins
Dokka is performing: documentation for debug-db-base
Validity check
Creating documentation models
Exiting Generation: Nothing to document


> Task :debug-db-base:dokkaJavadocJar
> Task :debug-db-base:generateMetadataFileForMavenPublication
> Task :debug-db-base:generatePomFileForMavenPublication
> Task :debug-db-base:signMavenPublication FAILED

57 actionable tasks: 56 executed, 1 up-to-date
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':debug-db-base:signMavenPublication'.
> Cannot perform signing task ':debug-db-base:signMavenPublication' because it has no configured signatory

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

BUILD FAILED in 1m 24s
```

**Comment by eygraber** [comment]
I'm seeing the issue on Gradle 6.8.2, 6.8.3, and 7.0

**Comment by akarnokd** [comment]
Same issue here. After upgrading to 0.15.0, I get "Could not find method signing() for arguments" on [this line](https://github.com/ReactiveX/RxJava/blob/3.x/build.gradle#L238). Is `apply plugin: 'signing'` a workaround or an intended change in requirements for the plugin?

**Comment by gabrielittner** [comment]
>I tried using signingInMemoryKey and now it fails when configuring the CloseAndReleaseRepositoryTask.

Weird, looking into it.

> Is apply plugin: 'signing' a workaround or an intended change in requirements for the plugin?

A workaround. It was an unintended side effect of another change (we are still applying the plugin ourselves but too late from your perspective). I will include a revert for that in 0.15.1

**Comment by gabrielittner** [comment]
@eygraber I missed the empty string in your snippet above `useInMemoryPgpKeys(signingKey, "")`. In that case you also need to set `signingInMemoryKeyPassword`/`ORG_GRADLE_PROJECT_signingInMemoryKeyPassword` with an empty value, otherwise the password we pass in is `null`. Looks like Gradle's internal error message for when it can't read the in memory key is the same as if you don't configure anything at all 😞

**Comment by eygraber** [comment]
Oh that's an unfortunate error message. I'll open an issue on Gradle.

**Comment by gabrielittner** [comment]
Thanks, I'll keep this open until the signing plugin being applied to late issue is fixed

From a quick test with our testing gpg key:
- this works `useInMemoryPgpKeys("lQdGBF4jUfwBEACblZV4uBViHcYLOb2280tEpr64iB9b6YRkWil3EODiiLd9JS3V7+BWpZVF8mbGy8AUR7T3GUsYmEvhHGw2s0IosOUVIu5W3eiU/K4CFMEsmV5JZLRSoHa+VJ3XOJA0ZQfBxwaLyq1vicgtOVVA1AJKctQZ3gfmw8u45NHBvtUOgAZQrDytKN1B/bcpxPqLOBi1RQuYOyjtLTGJoU+Jbf27H9EhXEJ80hWzd2c3khtZO3HkCbgHKIvEDi5qkb25zKluQbmMmhGpDyyR8XzIiVjPfSRG/+VXpWQlo9NjcaUWtaVyXYqN29z87BOdtD734b56kzHMhr1taDM89lkTKrgtuJBjZkA/wnDUG3+NZ1ly7Wc4z1sw7eovwrfBhDFpyWoASyoui8/78dZHflm3v9Xe+01d1mFeZzs9+UtlK9xUaGzpNBq6FYOKB8HaxgKKSNcFMio6XM9Wfmxc0boFE0fy6F9zIo9AQRB469ZPYO0NGxyBuAyihnd5NzAAHn5rFpYVNS0FKcQrZn939NE7y0in5dRdEfgRBsYeyoYhOf5XczNjHOma+ToMyvIPHMcaCcrGOeFBmsuY3EXa5MRA3cxudKK90dvXmv8bK+ZhFDRYq3S31ZtXAtwSjDTxj27XV8iqXI2Ne6We0oOROv1KY9v4CXg9XbPYYz6LHzLnj8lq/QARAQAB/gcDAnWMv+BoFys37xO+klZLzr9+JaJjXHP32h0lLii4TjKQerOUhAIaHLwv4AUpm+OFAEfozcNUX92wXkXm4QXNAKb+BQTDNGnd+iX/M1YVzBnp791XZo6k5gam/AzqznlrPgLHpx3kqSVR76Dj/KD3xxjy7iuklKywt1+lf5b2fFNhSGMyFqveCvmTg6jJGQ1DYEzZVYrG6VrWvHERIOUCSf8GlIEpk627rJoWC07ZRuOv0oHCJg+8D1D3Lc1vl3Wx49aLmWOOeHVaZPsk49I3OEAab/lVsP8q8bn6rjLzYpUx2TyFYmedTdEzx9SVG7VfozsHvV/n1BLiSphv3hD8cYmYRaiKvE65PHxiqa3rCsaZMgLhvmNXw81W8NN63A7wjqHk5u3CBU7FeymL5V5HsoePPv+jKWJtILSnt/NiLoLS8Wq5nBzL/J0hNo/HrMbK6TRjdMkfg0sIAB7+Y2A+cyOy9OMObgqFHSy5LbPXtkoO8axC/35DwfMiiTmCXFBnCAbiGniI+QWV+R9OR+sq0YrxUVL/4iIwq6FOZ4B7zjSMP9Q5n09S013TFrdg6yUw5wmDMzW9wadLxDJ/QB3G8rAwNvsWfPdn2EvaeV2blyPZwd+wmNvz+hQ7aUu5AUjlTA0ccsVYUCfIlL9aSOXOS42Go/xoTnU/ea9fo8IisbPSw2juebVcxExcetR84Mbc28qsfJ7I8vgtOEm87Zv+U2zPOZxje/QwRdoXxS/sLpfJ9GXRN/x9iLc2UJMXY92B/yD09sjgHhthkK/3Z0Tn92Lh56z3VmyxwRf4ZpSepogwZ9hnbP7h63Nm9yUxPek5N9zxp+o8DNw7t8gz5UAqSJthpxVpUE6a8VzzBTTy6yx3qUKHOuPWZEpoJvgwYybY/vzgsJUeBqRhWtHNEfArYQmbSXmaUgJPZ7dlxEx1bNz4pavI+AgsMUdcvtrNcnxdXcW+aEKwYW/CGJZJP4Vdue+FZACa5eoP1DpdtH7sU9Y2W78VcUef1/SNIHuS/HbIZpnHLK4XXIM3cyQxmmmmbazFDQwcKo4ckPXhB3PWAc/iocgmX+JXeSzSf91CWM/JcyWrh1qyzGdMw0+DrgWXFEJpI5U0VycycFdKwdHF0mF9FjaqpXmUlHJiJAk/TdaX3TJmhTfsW9QUbTbHk4Vrn/btyP6tZzxOZp0NIhJZVWKxgkA6XhHuvp9KUSZU6qjyTr7+CmQ1PH0FAm463Dw9LmGi8AQSUomItjR9zy77svGHtcXGKSSFm7FRTEOfEN500czgCVYKTpnEzDWLur+Mc7H/A5ycgP/7GPthgHqOmW5m2Ozibo8HBtVkbXUARbIrdVyF/Haqy6+k3beiUTP/JKdb7x35AX6d3cS8WhPumQIleAT+as9Pmc1po8dkKOXtQeio21jdBU5R8xiw/PrM6tcBo829alWpDiTBUgFy8oHJNIYqjz9f01OorObz9MkVDdsPbR8KjdAcqAm/JR9IRkZnr9q7lUgitYzrj3yvOZwIMxmBdhEyYO35xgQsdx0f/uIfskWwRKu4YK3aCTA1/sORPzL+qdqarxJVgx2ukVBYf599tA4ybmu58b5KtshEb9yPTlxhLiUyWG25asyaE8Po8AfXhhDAiljrqSgZZ6Kcs2jjx4DFqS3wHJQoETH9wtlpFvyWs+85hWcO7YCjYGT6wCbO9sa77Gz1QLg5BqOjnkdoOgCnrOip8g9Noj5FX5aK6VTk5eX3CjQ+CLtReyx4gcaL+oEA8Sj83RFzv/4rSZCs8MW0JnZhbm5pa3RlY2gvZ3JhZGxlLW1hdmVuLXB1Ymxpc2gtcGx1Z2luiQJUBBMBCgA+AhsDBQsJCAcDBRUKCQgLBRYCAwEAAh4BAheAFiEE93Y2HTpkSt27r5YoNnBrHbicQFUFAmBN9W8FCQuQpPMACgkQNnBrHbicQFXoVA/+PVxJfyyC58RT3y6oIiuThYk3vNnCD0OMb4SzsxHgTyhOyvCS3UJSUsxFkSasPyCARG4gANeTcvmVwK66xUn9piYVaitjBMaD6VZWXORGs3/9MGdOF7YVvLuhxq6dR3tsS97j3AvM5D5+6HF4MEJePVAe6AcH8xDFWQMsU1KntysJ6xjG8kARGzY5ipD7bp2hQ1ZbNhkXKTqrJ0ZJoPfv3EK/Gfa79ZqtKbxXlTHf8k2fDVXEEPyYJGOtTR1B2V/xhAtGmPhDTJgzESgEtEplPjwZ+KydVBZq5dP9DBIzHmgUDDM2MmILyFXnbZsiVUpiFTNM+nKqhlpxbWahNPuqyrRq5zPoxO3jHPFY/pcgZxlHE/y3S5liMoTdxlbJoH2iDHp8WUXXvyONnP5R0TTuHOkUi0uj+gqjRwRjR+pFtVSIV7HQN+mu6m1jFuDo2eR9bTROxVSH30XE3IOlfxTeqMuUo6PFc/50X6L7xTODnQ1MEoDDmtIckHEa4oQlRHCN7pfVNcCg4LC40MIw9PGE6VRLpfWKkI8OdTq7byJhyA4qFEVshy6lyoLnstQbQIv0xfJmBaP9NWsrdZ2xYBVuvUumswHfSLUEqadKznzx0u19tljGyCyYdueDtU5oXs9f9WwkQzi8oAcjZfGWlkvOPhcriPC0mrMP8ZqeKBBMhamJAlQEEwEKAD4WIQT3djYdOmRK3buvlig2cGsduJxAVQUCXiNR/AIbAwUJlmQMgAULCQgHAwUVCgkICwUWAgMBAAIeAQIXgAAKCRA2cGsduJxAVaUYD/0S/SdLsEr0h3jgRvl8O8KC/ncGSlvkW6d02tQZz5LrRC/yefBEAmmxugi3oCU168aGq/L/J0aXh7XunizT3oKaktFUsYNRH4InZ3R7yP1+MgPi3PsK/6cM1q/Rw5nIu5lsq/v05LzF5yhXSPvdrK+TqXsjaBvpDvHdwRqVpxkcT7jpS2ZOc4zEXyf6ibnzoePf9DfrKRJ4ZHp844pJwuOmxpNl6Gd7zkWeI5b7vvLI+hGw1IIb9l5oOF60/g1RrGrIcBHrzcW+fXfeil44JMIh3k4UnmqGSezAANoGIJkrhmpioRR1M/CuiL7ON8xg6JHC3M7CZjnP7rntiJI8Em6pP5bSxGOIQ/ZgcI54UY4tu2m7T5DpszQAQKht18UswzgVpItkrLHTmTNzRbWac27PmJPCc8Uhn2XIbSH5/TjGkNTEJblbQtAa5Hz+1bI+eUH5wNwkkHuaViqWq6IIE/gZ23WhU6JlUllNez83MXaImHxLTMdLL5UG1OPL7M3ze+H8rWVg+rB4+MujeCbYXgnyfsjYMErqnrnpxyChaCnQjyvh8JYoYr56prE62pnSU7e5MRtPGNTOaGCdY231Q/AqqdcjJdfpt0KqEv4bkkJe/SVgQgoXY3JCPB35MUBsu7h8uRIfb+Lj7WfjwS0/h1v3tekJCYz0dDSQoCPTCOrgy50HRgReI1H8ARAA3A5OqE/B5s0LZZzNFfcuaLW2mgD6sFOibu2oVvas7u1jllmgK1xL8Iz0vsO9gQa10B0IK8Lgm5RnkW3WqbBFP+Cn/ZLhWjJ6sZIuko6oT5A/wvOozK8w52Er9MNC7VEvc+gbBxS1pyVNd+4J+OLQkBDkkCTivivruChCsaKM+XCzpf9Mh+5u+ZN26rHXCzeGOCIAxhIXKHGPPNXZ7CId7McttDsBKYMBxCj+q9gkd4eazdbkbNWYHm9pkAFMwyTQqNpPo9xrY7MhZMjhGaXB88vkT0fnwFObaAQvuklIen1K9ojrTG5xDDz0iJP0QqjJplOyJsPi/aPo0hmEsB7lPwMpv9nq/b9NBlixzdSpn6s0YW3ufdHaH3JnYoGX6dzKNEPE013jf8ZzEIqanSPieKbqSG4HnxEhfQgl13D0JFEqv1EmPbi6APMM+Pj9cI/mFxJOG8TBzWf55mIPxk/s9IOSgsEvhbJclRbkDFKRS9xNkY7lweGqbZd8hGOIt4Yo1T2Bz/PZLmz5jReh10kFwhbSJai2LpQEH7BhpomDIXC4YYxBR/YnFy6HwXIaeB+uw7NnSuV3E7R/a0gY1UpLJJKlIY3PYv0G13AfK6mnRjOOWQYo5Y+wOJr/J8eJer4Vma9NOb0krVjKimwudev7cIYa0A/5q1UdEd0R/an27nUAEQEAAf4HAwIJRkyXMxklr+9Tbu4swUl88SERNwrbAFYorvDuG4n0B3jhicAuN5rOoHFLZ87Y0hcZFeR1Yr0SSgfBeEA0wWQQXKrmQmjJADhzJRPfPndMWlpqikWQONqYY3TDquxOQWrcF4H/2ylySNUitYOEWNQiHKD5f8MJFzTAJRphQzS24a3dsk+Y/2KIjUyNgf8vmLLx5BdXoOcehR6874fmFnKNPuuFKoU+3aY471LYNpQm7Upmnkk7IPv7OTqIF9irCJwK1j81J8pOi1NkDGEKirX40+BffurSu3nSjwTyvG/5l1jaLPG6w8nMVPShxPBIcD4kZ823tUwjndPHjJGW5RgTmuLSgAP+lX0GY2f5DIhEcT363ZsuaIoonHQ3pRtQlspz9x3whl8MieanUlrLIuvLvPHB/bJRGr5quelWqL1Pa1iG3ymQpSw2HaN1HBk2koD3egCN6LBaQMi8g3Nl5pZvGe4Fus9Aps8GXb0Mmv/J65CHFpYrmodgUe3HxNCxkPUXgDlywi3t9UUnX4CbXhHAxsnqLCF5rJqeThfLOFbqrkYu4/PcMrVU6vz+obnVmXuRUHBPM4nMoBzeyOBXeuyq8BmhbkiDCbHs77Gv04EefGl72lkROIJINeaWTc3T4YrQCRMpS3CBZIcrj9bJsRRkZXsm4FjL4Z+7HP7ItCf++cZY53aeY7XXzrlBBaAT9vdEhyjRk4uaRjtz9zJp2Vn/wJiZ9Od/yPEw61nhKwQTuoEE1xQIHHTqAyO8C0mVucRg+tqSQwijFTZcXu7EZn8Xqs+bnqLFdcBzh5NJyBicepe0IF4Q5wrSUOy8ZDAUvPRlsYFPLuJn4nTKmEKOYxcjVh4riI+u0arAzINJBqyERPtrgqhJm++iPSvu5cP54873QH2Lnqx0PqefomXhB7XAVLL4FC+rX9L7Ozwu6VXaOu+TAXr0phSXagSvwgD6ZVXOgtGSHXgNjD/w447Ptn8cEA0GbZUFr2UAI+CN4K05ifqUspnEgDlx0jeJ7K4Snq+K9rOxRgOnWgGZOZEhW4y2had31ULbaThnIpxAxBANu/1XemLvH8rVWkdolT9iU2ST9H5c/1z+PN3BMqCmmE0blDNLE3TrZplw4VA9PIoTocvv7IhGcUdqIYN2MysP8dFUA3tG0I9XmKO1hIT2RYXJ7h5ffBr4lwxdBKcU27T6SJoPtfDs3QCoeB9IM8QotQgZc0kUpAhSvFPhIu7MbpWmxwU5CsZNYQOYkTOa6H1mWhmephgzzNbdjaFgLmoX4Q/VCsobMGVA9wN6Wi7oMKSVtFwUPhnYLVfbykkIy/AxldYcTmi2wD0BFKwxuQkmVgr76PX6WeucU9o9PTiHAQaI90ahlaOlWdjJrrd74b9vyn1eGC2FVz1J+UsXjEvj8koTLXEKXNQKfbI2H5a9+kxrpdaKvoLRlP+DJkVbSeerRxhROXBQxyZ6hkTyVKSFrJtbbmWLBC5vIBzTlhq91wU3BrbsCKrd++GJrS19ePoD0vvhim+9hPlHc1Ah75AfxJ7uCgcXxt+WNNs2ZYnDfgDfpzNXVztmsR7iph1N9NI2nPWv6+kmHOXj6nt5kL7CedX5qTEusmPC54LVgvpKwW4hOtGvXiVzEYqVX/Jw+Ha81DhAEMWr3GMK0ZkIrAXI6rbKGh/KlrJC6FfNqf/Zq5Z6FGPSE8WLH8o5ji77lCl2ppD2vacvd0WQ4dlQc+HdgFA/9jRO9JtxxrmNx8cmedrffaJPv06iRpmVmmsrUE3DcqqROVVciQI8BBgBCgAmFiEE93Y2HTpkSt27r5YoNnBrHbicQFUFAl4jUfwCGwwFCZZkDIAACgkQNnBrHbicQFX0+g/8DzgLu0lJf+oljb9i6kx1wIxLR0Juhj+zyZtoKyEmzxshQENbz5z4hlDSqmhozs7Ds9Tq2ABfYA6nDO3mvizNPtadKH8kbnbXhMmemuxbIAD7UfvmhNRaXKiNyl6KZvSHVs/AQGcWdqjNnwphTiLmnrAhX+aGM2nxPu9AgWEE2AgddrOHLWX/vadC6+yR8GrMdAgPTfsCWOLipsZ0kBjxgHrcqABubvNQfibgzJb9vJh7V/4zvcCZz3yQ+zZ+XtqoFfNR9LfloeKsMux5ZaSHPi4iZfCs7tJioV7mzUDWe6IM2XlgK/XgoyxIPdikiWam5aDNLb2+E+sU3o52EdTL/yBg7tdaWBf42GoGC8di/rSmpul89VHtejoDAzJlUtw9NE7wmMRhxJPriW32PbIvcMlsk0JFlQpPBKK1Kkxttni68IWxuLKrsthXEqSerqV86L7fdtEtXONnjKxRPgT6omSbgDsBZjn8Om8h8fC5ZmZNDp6Cfchdp1oKANA7vimtuplILcGmqQ7bfvUmf2v+PbE4xcBxGO5UM4Il2s30KxZh5v89frKX7i8bXuQUOIGQI+X7zyVxtLmIdzWZe5Z4+Vb5D/q03nk7LqZfwn90YtUFgFApuYKEa5GVl3BYmAeH47ms3loMAR1r9pip+B1QLwEdLCEJA+3IIiw4qM5hnMw=", "test")`
- using an incorrect password `useInMemoryPgpKeys("....", "abc")` gives this error message `> Could not read PGP secret key`
- however passing null as password `useInMemoryPgpKeys("....", null)` which is what happened in this case gives `> Cannot perform signing task ':signMavenPublishBasePluginPluginMarkerMavenPublication' because it has no configured signatory`

**Comment by eygraber** [comment]
I have no problem adding the extra env var, but are you able to detect a null password and use an empty string instead as part of this plugin?

**Comment by eygraber** [comment]
https://github.com/gradle/gradle/issues/16983

**Comment by gabrielittner** [comment]
This should be fixed in 0.15.1 which was just released
